expression/builtin_string.go

```diff
@@ -3260,7 +3277,12 @@ func (b *builtinInsertSig) evalString(row chunk.Row) (string, bool, error) {
 	}
 
 	if length > runeLength-pos+1 || length < 0 {
-		return string(runes[0:pos-1]) + newstr, false, nil
+		length = runeLength - pos + 1
+	}
+
+	if uint64(runeLength-length)*uint64(mysql.MaxBytesOfCharacter)+uint64(len(newstr)) > b.maxAllowedPacket {
```

Comment:
What if the number of bytes of some character is less than `MaxBytesOfCharacter`? Will this raise warning while it should not?

Reply:
Yes, this false-negative case exists, but to compute the exact bytes of these characters, we may need the charset info of the string, and then compute the length of specific characters for this charset, which incurs too much overhead, I guess that is why we use `MaxBytesOfCharacter` in other functions such as `builtinLpadSig::evalString`. @zz-jason should we compute the exact bytes for the string?

Reply:
maybe we can build the string first, and use len() to compute the consumed bytes, but is it possible that panic is raised in building the string because it is too large? I think the purpose of `max_allowed_packet` check here is to prevent this kind of panic to some extent.

Reply:
in MySQL, the byte count for `lpad` is calculated by:
```cpp
byte_count = count * collation.collation->mbmaxlen;
```

as for the `insert` function, seems MySQL calculates the exact bytes for the string:
```cpp
  /*
    There is one exception not handled (intentionaly) by the character set
    aggregation code. If one string is strong side and is binary, and
    another one is weak side and is a multi-byte character string,
    then we need to operate on the second string in terms on bytes when
    calling ::numchars() and ::charpos(), rather than in terms of characters.
    Lets substitute its character set to binary.
  */
  if (collation.collation == &my_charset_bin) {
    res->set_charset(&my_charset_bin);
    res2->set_charset(&my_charset_bin);
  }

  /* start and length are now sufficiently valid to pass to charpos function */
  start = res->charpos((int)start);
  length = res->charpos((int)length, (uint32)start);
```

Reply:
Why first build it? It just constitutes of three parts, so sum them would be enough.

Reply:
seems there isn't a convenient and efficient way to calculate the byte count of a `[]rune`. @tiancaiamao any idea?

Reply:
another method is to write a utf-8 code point iterator like this: https://gist.github.com/zz-jason/078110974bb931b7f8e3432775ecfd05

Reply:
oh you are right... updated